clippy_lints/src/manual_ignore_case_cmp.rs

```diff
@@ -0,0 +1,98 @@
+    /// memory for the new strings, and it is more readable.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// fn compare(a: &str, b: &str) -> bool {
+    ///     a.to_ascii_lowercase() == b.to_ascii_lowercase()
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// fn compare(a: &str, b: &str) -> bool {
+    ///    a.eq_ignore_ascii_case(b)
+    /// }
+    /// ```
+    #[clippy::version = "1.82.0"]
+    pub MANUAL_IGNORE_CASE_CMP,
+    nursery,
+    "manual case-insensitive ASCII comparison"
+}
+
+declare_lint_pass!(ManualIgnoreCaseCmp => [MANUAL_IGNORE_CASE_CMP]);
+
+fn get_ascii_type<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'_>) -> Option<Ty<'tcx>> {
+    let ty_raw = cx.typeck_results().expr_ty(expr);
+    let ty = ty_raw.peel_refs();
+    if needs_ref_to_cmp(cx, ty)
+        || ty.is_str()
+        || ty.is_slice()
+        // FIXME: Need a better way to check if ty is OsStr(ing)
+        || ["std::ffi::OsStr", "std::ffi::OsString"].contains(&ty.to_string().as_str())
+    {
+        Some(ty_raw)
+    } else {
+        None
+    }
+}
+
+/// Returns true if the type needs to be dereferenced to be compared
+fn needs_ref_to_cmp(cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
+    ty.is_char()
+        || *ty.kind() == ty::Uint(UintTy::U8)
+        || is_type_diagnostic_item(cx, ty, sym::Vec)
+        || is_type_lang_item(cx, ty, LangItem::String)
+}
+
+impl LateLintPass<'_> for ManualIgnoreCaseCmp {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
+        // check if expression represents a comparison of two strings
+        // using .to_ascii_lowercase() or .to_ascii_uppercase() methods
+        // Offer to replace it with .eq_ignore_ascii_case() method
+        if let Binary(op, left, right) = &expr.kind
+            && (op.node == BinOpKind::Eq || op.node == BinOpKind::Ne)
+            && let MethodCall(left_path, left_val, _, _) = left.kind
+            && let MethodCall(right_path, right_val, _, _) = right.kind
+            && left_path.ident == right_path.ident
+            && matches!(
+                left_path.ident.name.as_str(),
+                "to_ascii_lowercase" | "to_ascii_uppercase"
+            )
+            && get_ascii_type(cx, left_val).is_some()
+            && let Some(rtype) = get_ascii_type(cx, right_val)
+        {
+            // FIXME: there must be a better way to add dereference operator
+            let deref = if needs_ref_to_cmp(cx, rtype) { "&" } else { "" };
+            span_lint_and_sugg(
+                cx,
+                MANUAL_IGNORE_CASE_CMP,
+                expr.span,
+                "manual case-insensitive ASCII comparison",
+                "consider using `.eq_ignore_ascii_case()` instead",
+                format!(
+                    "{}.eq_ignore_ascii_case({deref}{})",
+                    snippet(cx, left_val.span, "_"),
+                    snippet(cx, right_val.span, "_")
```

Comment:
This should use `snippet_with_applicability(..)` to correctly update the applicability if `snippet` fails.

Reply:
fixed